File: app/src/main/java/feelings/guide/ui/log/AnswerLogActivity.java
package feelings.guide.ui.log;

import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.TextView;

import androidx.annotation.Nullable;
import androidx.fragment.app.DialogFragment;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.ItemTouchHelper;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.snackbar.Snackbar;

import feelings.guide.R;
import feelings.guide.answer.Answer;
import feelings.guide.answer.AnswerStore;
import feelings.guide.question.QuestionService;
import feelings.guide.ui.BaseActivity;
import feelings.guide.ui.answer.AnswerActivity;
import feelings.guide.ui.question.QuestionClearLogDialogFragment;
import feelings.guide.util.ToastUtil;

import static feelings.guide.FeelingsApplication.QUESTION_ID_PARAM;
import static feelings.guide.ui.answer.AnswerActivity.ANSWER_ID_PARAM;

public class AnswerLogActivity extends BaseActivity implements
        ClearLogFullDialogFragment.ClearLogFullDialogListener,
        ClearLogDeletedDialogFragment.ClearLogDeletedDialogListener,
        QuestionClearLogDialogFragment.QuestionClearLogDialogListener,
        AnswerLogSwipeCallback.AnswerLogSwipeListener {

    private static final long DEFAULT_QUESTION_ID = -1;
    public static final String REFRESH_ANSWER_LOG_KEY = "should-refresh-answer-log";
    private static final int UPDATE_ANSWER_REQUEST_CODE = 101;
    private long questionId;
    private boolean isFull;
    private RecyclerView recyclerView;
    private AnswerLogAdapter adapter;
    private Answer lastDeleted;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        questionId = getIntent().getLongExtra(QUESTION_ID_PARAM, DEFAULT_QUESTION_ID);
        isFull = questionId == DEFAULT_QUESTION_ID;

        setContentView(isFull
                ? R.layout.full_answer_log_activity
                : R.layout.answer_log_by_question_activity);

        recyclerView = findViewById(R.id.answer_log_recycler_view);
        TextView emptyLogText = findViewById(R.id.answer_log_empty);

        adapter = new AnswerLogAdapter(this, isFull, questionId);
        boolean notEmpty = adapter.isNotEmpty();
        recyclerView.setVisibility(notEmpty ? View.VISIBLE : View.GONE);
        emptyLogText.setVisibility(notEmpty ? View.GONE : View.VISIBLE);
        if (notEmpty) {
            setUpRecyclerView();
        }
        fillQuestionText();
    }

    private void setUpRecyclerView() {
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(layoutManager);
        DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(recyclerView.getContext(),
                layoutManager.getOrientation());
        recyclerView.addItemDecoration(dividerItemDecoration);

        recyclerView.setAdapter(adapter);

        ItemTouchHelper itemTouchHelper = new ItemTouchHelper(new AnswerLogSwipeCallback(this, this));
        itemTouchHelper.attachToRecyclerView(recyclerView);
    }

    private void fillQuestionText() {
        if (!isFull) {
            TextView questionText = findViewById(R.id.question_text_on_answer_log);
            questionText.setText(QuestionService.getQuestionText(this, questionId));
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (adapter != null) {
            adapter.changeCursor(null);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        if (adapter.isNotEmpty()) {
            MenuInflater inflater = getMenuInflater();
            inflater.inflate(isFull
                            ? R.menu.answer_log_full_menu
                            : R.menu.answer_log_by_question_menu,
                    menu);
        }
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.clear_log_full:
                showClearLogFullConfirmation();
                return true;
            case R.id.clear_log_deleted:
                showClearLogDeletedConfirmation();
                return true;
            case R.id.clear_log_by_question:
                showClearLogByQuestionConfirmation();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void showClearLogFullConfirmation() {
        DialogFragment dialogFragment = new ClearLogFullDialogFragment();
        dialogFragment.show(getSupportFragmentManager(), ClearLogFullDialogFragment.class.getSimpleName());
    }

    private void showClearLogDeletedConfirmation() {
        DialogFragment dialogFragment = new ClearLogDeletedDialogFragment();
        dialogFragment.show(getSupportFragmentManager(), ClearLogDeletedDialogFragment.class.getSimpleName());
    }

    private void showClearLogByQuestionConfirmation() {
        DialogFragment dialogFragment = new QuestionClearLogDialogFragment();
        dialogFragment.show(getSupportFragmentManager(), QuestionClearLogDialogFragment.class.getSimpleName());
    }

    @Override
    public void onClearLogFullConfirmed() {
        AnswerStore.deleteAll(this);
        ToastUtil.showShort(this, getString(R.string.msg_clear_log_full_success));
        adapter.refresh();
    }

    @Override
    public void onClearLogDeletedConfirmed() {
        AnswerStore.deleteForDeletedQuestions(this);
        ToastUtil.showShort(this, getString(R.string.msg_clear_log_deleted_success));
        adapter.refresh();
    }

    @Override
    public void onClearLogByQuestionConfirmed() {
        AnswerStore.deleteByQuestionId(this, questionId);
        ToastUtil.showShort(this, getString(R.string.msg_clear_log_by_question_success));
        adapter.refresh();
    }

    @Override
    public void onDeleteAnswer(int position) {
        lastDeleted = adapter.getByPosition(position);
        AnswerStore.deleteById(this, lastDeleted.getId());
        //notifying adapter only about 1 position change does not work for some reason, so update everything as a workaround
        adapter.refresh();
        showUndoDeleteSnackbar();
    }

    private void showUndoDeleteSnackbar() {
        View view = findViewById(R.id.answer_log_view);
        Snackbar snackbar = Snackbar.make(view, R.string.msg_answer_deleted_success, Snackbar.LENGTH_LONG);
        snackbar.setAction(R.string.snackbar_undo, v -> undoDelete());
        snackbar.show();
    }

    private void undoDelete() {
        AnswerStore.saveAnswer(this, lastDeleted);
        adapter.refresh();
    }

    @Override
    public void onEditAnswer(int position) {
        Answer answer = adapter.getByPosition(position);
        Intent intent = new Intent(this, AnswerActivity.class);
        intent.putExtra(QUESTION_ID_PARAM, answer.getQuestionId());
        intent.putExtra(ANSWER_ID_PARAM, answer.getId());
        startActivityForResult(intent, UPDATE_ANSWER_REQUEST_CODE);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == UPDATE_ANSWER_REQUEST_CODE && resultCode == RESULT_OK) {
            boolean refresh = data != null && data.getBooleanExtra(REFRESH_ANSWER_LOG_KEY, false);
            if (refresh) {
                adapter.refresh();
                //redraw recyclerView
                recyclerView.setAdapter(adapter);
            }
        }
    }
}

File: app/src/main/kotlin/feelings/guide/ui/log/AnswerLogActivity.kt
package feelings.guide.ui.log

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import com.google.android.material.snackbar.Snackbar
import feelings.guide.ANSWER_ID_PARAM
import feelings.guide.ANSWER_IS_ADDED_OR_UPDATED_RESULT_KEY
import feelings.guide.EDIT_ANSWER_REQUEST_CODE
import feelings.guide.EXPORT_LOG_REQUEST_CODE
import feelings.guide.QUESTION_ID_PARAM
import feelings.guide.R
import feelings.guide.UPDATED_ANSWER_ID_RESULT_KEY
import feelings.guide.answer.Answer
import feelings.guide.answer.AnswerStore
import feelings.guide.databinding.AnswerLogHostBinding
import feelings.guide.export.LogExporter
import feelings.guide.ui.BaseActivity
import feelings.guide.ui.answer.AnswerActivity
import feelings.guide.ui.log.byquestion.AnswerLogByQuestionFragment
import feelings.guide.ui.log.full.AnswerLogFullFragment
import feelings.guide.util.EXPORT_FILE_NAME_FORMATTER
import java.time.LocalDateTime

const val DEFAULT_QUESTION_ID: Long = -1

class AnswerLogActivity : BaseActivity() {
    private var isFull: Boolean = false
    private var questionId: Long = DEFAULT_QUESTION_ID
    private lateinit var binding: AnswerLogHostBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = AnswerLogHostBinding.inflate(layoutInflater)
        setContentView(binding.root)

        if (savedInstanceState == null) {
            questionId = intent.getLongExtra(QUESTION_ID_PARAM, DEFAULT_QUESTION_ID)

            isFull = questionId == DEFAULT_QUESTION_ID
            val fragment = when {
                isFull -> AnswerLogFullFragment(this::createFileForExport)
                else -> AnswerLogByQuestionFragment(questionId, this::createFileForExport)
            }
            supportFragmentManager
                .beginTransaction()
                .add(R.id.answerLogContent, fragment)
                .commit()
        }
    }

    internal fun navigateToEditAnswer(answer: Answer) {
        startActivityForResult(Intent(this, AnswerActivity::class.java).apply {
            putExtra(QUESTION_ID_PARAM, answer.questionId)
            putExtra(ANSWER_ID_PARAM, answer.id)
        }, EDIT_ANSWER_REQUEST_CODE)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == EDIT_ANSWER_REQUEST_CODE && resultCode == RESULT_OK) {
            // returned from editing the answer
            val answerLogFragment = supportFragmentManager.findFragmentById(R.id.answerLogContent)
            val answerId = data?.getLongExtra(UPDATED_ANSWER_ID_RESULT_KEY, -1) ?: -1
            val answerIsUpdated =
                data?.getBooleanExtra(ANSWER_IS_ADDED_OR_UPDATED_RESULT_KEY, false)
                    ?: false
            if (isFull)
                (answerLogFragment as? AnswerLogFullFragment)?.onReturnFromEditAnswer(
                    answerId,
                    answerIsUpdated
                )
            else
                (answerLogFragment as? AnswerLogByQuestionFragment)?.onReturnFromEditAnswer(
                    answerId,
                    answerIsUpdated
                )
        } else if (requestCode == EXPORT_LOG_REQUEST_CODE && resultCode == RESULT_OK) {
            try {
                if (data?.data == null) {
                    Snackbar.make(
                        binding.answerLogContent, "Choose file location to export.",
                        Snackbar.LENGTH_LONG
                    ).show()
                }
                data?.data?.let { uri: Uri ->
                    val answers = AnswerStore.getAnswersForExport(this, questionId)
                    contentResolver.openOutputStream(uri).use {
                        if (it == null) {
                            Snackbar.make(
                                binding.answerLogContent, "Cannot write to file $uri",
                                Snackbar.LENGTH_LONG
                            ).show()
                            return
                        }

                        LogExporter().export(answers, it, this)

                        Snackbar.make(
                            binding.answerLogContent, "Finished exporting to file",
                            Snackbar.LENGTH_LONG
                        ).show()
                    }
                }
            } catch (e: Exception) {
                Snackbar.make(
                    binding.answerLogContent, "An error occurred while exporting to file. ${e.message}",
                    Snackbar.LENGTH_LONG
                ).show()
                Log.e("export", "An error occurred while creating a file.", e)
            }
        }
    }

    private fun createFileForExport() {
        try {
            val id = if (isFull) "full" else questionId.toString()
            val timestamp = LocalDateTime.now().format(EXPORT_FILE_NAME_FORMATTER)
            val fileName = "feelings-guide-log-$id-$timestamp.csv"

            val intent = Intent(Intent.ACTION_CREATE_DOCUMENT).apply {
                addCategory(Intent.CATEGORY_OPENABLE)
                type = "text/csv"
                putExtra(Intent.EXTRA_TITLE, fileName)
            }
            startActivityForResult(intent, EXPORT_LOG_REQUEST_CODE)
        } catch (e: Exception) {
            Snackbar.make(
                binding.answerLogContent, "An error occurred while creating a file. ${e.message}",
                Snackbar.LENGTH_LONG
            ).show()
            Log.e("export", "An error occurred while creating a file.", e)
        }
    }
}
